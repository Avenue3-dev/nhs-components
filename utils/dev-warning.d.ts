/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
