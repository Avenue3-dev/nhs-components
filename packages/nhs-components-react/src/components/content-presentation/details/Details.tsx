import clsx from 'clsx'
import { HTMLProps } from 'react'

type DetailsProps = {
  expander?: boolean
} & HTMLProps<HTMLDetailsElement>

type Details = {
  Summary: React.FC<HTMLProps<HTMLDivElement>>
  Text: React.FC<HTMLProps<HTMLDivElement>>
  ExpanderGroup: React.FC<HTMLProps<HTMLDivElement>>
} & React.FC<DetailsProps>

/**
 * A component to display within a Details component.
 *
 * @param {HTMLProps<HTMLDivElement>} props
 * @param {string} [props.className] - An optional class name to add to the component.
 *
 * @example
 * ```tsx
 * <Details.Summary>
 *  How to find your NHS number
 * </Details.Summary>
 * ```
 */
const Summary: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <summary className={clsx('nhsuk-details__summary', className)} {...rest}>
      <span className="nhsuk-details__summary-text">{children}</span>
    </summary>
  )
}

/**
 * A component to display within a Details component.
 *
 * @param {HTMLProps<HTMLDivElement>} props
 * @param {string} [props.className] - An optional class name to add to the component.
 *
 * @example
 * ```tsx
 * <Details.Text>
 *  You can find your NHS number...
 * </Details.Text>
 */
const Text: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <div className={clsx('nhsuk-details__text', className)} {...rest}>
      {children}
    </div>
  )
}

/**
 * A container for multiple expanders.
 *
 * @param {HTMLProps<HTMLDivElement>} props
 * @param {string} [props.className] - An optional class name to add to the component.
 *
 * @example
 * ```tsx
 * <Details.ExpanderGroup>
 *  <Details expander>
 *    <Details.Summary>Question 1</Details.Summary>
 *    <Details.Text>Answer 1</Details.Text>
 *  </Details>
 *  <Details expander>
 *    <Details.Summary>Question 2</Details.Summary>
 *    <Details.Text>Answer 2</Details.Text>
 *  </Details>
 * </Details.ExpanderGroup>
 * ```
 */
const ExpanderGroup: React.FC<HTMLProps<HTMLDivElement>> = ({
  className,
  children,
  ...rest
}): JSX.Element => {
  return (
    <div className={clsx('nhsuk-expander-group', className)} {...rest}>
      {children}
    </div>
  )
}

/**
 * Make a page easier to scan by letting users reveal more detailed information only if they need it.
 *
 * For more information on when to use this component, go the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/details).
 *
 * @param {boolean} [props.expander] - If true, the component will be styled as an expander.
 * @param {string} [props.className] - An optional class name to add to the component.
 *
 * @example
 *
 * ```tsx
 * <Details>
 *  <Details.Summary>How to find your NHS number</Details.Summary>
 *  <Details.Text>You can find your NHS number...</Details.Text>
 * </Details>
 * ```
 */
const Details: Details = ({
  children,
  expander,
  className,
  ...rest
}): JSX.Element => {
  return (
    <details
      className={clsx(
        'nhsuk-details',
        { 'nhsuk-expander': expander },
        className,
      )}
      {...rest}
    >
      {children}
    </details>
  )
}

Details.displayName = 'Details'
Summary.displayName = 'Details.Summary'
Text.displayName = 'Details.Text'
ExpanderGroup.displayName = 'Details.ExpanderGroup'

Details.Summary = Summary
Details.Text = Text
Details.ExpanderGroup = ExpanderGroup

export default Details
