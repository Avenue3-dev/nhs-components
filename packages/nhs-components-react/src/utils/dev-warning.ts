const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
