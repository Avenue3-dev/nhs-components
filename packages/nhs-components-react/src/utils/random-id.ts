export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
