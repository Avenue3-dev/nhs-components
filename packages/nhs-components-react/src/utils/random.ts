export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
