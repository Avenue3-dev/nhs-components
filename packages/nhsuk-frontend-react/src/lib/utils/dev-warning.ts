const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
