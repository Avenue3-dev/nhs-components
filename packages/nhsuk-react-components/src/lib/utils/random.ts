export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
