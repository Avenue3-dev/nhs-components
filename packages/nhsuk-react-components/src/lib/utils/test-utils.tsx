/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
