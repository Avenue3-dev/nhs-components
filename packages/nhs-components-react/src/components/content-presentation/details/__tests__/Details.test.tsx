import { render, screen } from '@/utils/test-utils'
import { describe } from 'vitest'
import Details from '..'

describe('Details', () => {
  it('should match snapshot', () => {
    render(
      <Details data-testid="details">
        <Details.Summary>Summary text</Details.Summary>
        <Details.Text>Details text</Details.Text>
      </Details>,
    )

    expect(screen.getByTestId('details')).toMatchSnapshot()
  })

  it('should match snapshot - expander', () => {
    render(
      <Details data-testid="details" expander>
        <Details.Summary>Summary text</Details.Summary>
        <Details.Text>Details text</Details.Text>
      </Details>,
    )

    expect(screen.getByTestId('details')).toMatchSnapshot()
  })

  it('should match snapshot - expander group', () => {
    render(
      <Details.ExpanderGroup>
        <Details data-testid="details" expander>
          <Details.Summary>Summary text</Details.Summary>
          <Details.Text>Details text</Details.Text>
        </Details>
      </Details.ExpanderGroup>,
    )

    expect(screen.getByTestId('details')).toMatchSnapshot()
  })
})
