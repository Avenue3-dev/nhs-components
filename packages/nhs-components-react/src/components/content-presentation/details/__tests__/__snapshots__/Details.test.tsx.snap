// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Details > should match snapshot - expander 1`] = `
<details
  class="nhsuk-details nhsuk-expander"
  data-testid="details"
>
  <summary
    class="nhsuk-details__summary"
  >
    <span
      class="nhsuk-details__summary-text"
    >
      Summary text
    </span>
  </summary>
  <div
    class="nhsuk-details__text"
  >
    Details text
  </div>
</details>
`;

exports[`Details > should match snapshot - expander group 1`] = `
<details
  class="nhsuk-details nhsuk-expander"
  data-testid="details"
>
  <summary
    class="nhsuk-details__summary"
  >
    <span
      class="nhsuk-details__summary-text"
    >
      Summary text
    </span>
  </summary>
  <div
    class="nhsuk-details__text"
  >
    Details text
  </div>
</details>
`;

exports[`Details > should match snapshot 1`] = `
<details
  class="nhsuk-details"
  data-testid="details"
>
  <summary
    class="nhsuk-details__summary"
  >
    <span
      class="nhsuk-details__summary-text"
    >
      Summary text
    </span>
  </summary>
  <div
    class="nhsuk-details__text"
  >
    Details text
  </div>
</details>
`;
